shared/teams/channel/index.tsx

```diff
@@ -4,67 +4,210 @@ import * as Types from '../../constants/types/teams'
+export type OwnProps = Container.RouteProps<{
   teamID: Types.TeamID
   conversationIDKey: ChatTypes.ConversationIDKey
-  sections: Sections
+  selectedTab?: TabKey
+}>
+
+const useLoadDataForChannelPage = (
+  teamID: Types.TeamID,
+  conversationIDKey: ChatTypes.ConversationIDKey,
+  selectedTab: TabKey,
+  meta: ChatTypes.ConversationMeta,
+  participants: string[],
+  bots: string[]
+) => {
+  const dispatch = Container.useDispatch()
+  const prevSelectedTab = Container.usePrevious(selectedTab)
+  const featuredBotsMap = Container.useSelector(state => state.chat2.featuredBotsMap)
+  React.useEffect(() => {
+    if (selectedTab !== prevSelectedTab && selectedTab === 'members') {
+      if (meta.conversationIDKey === 'EMPTY') {
+        dispatch(
+          Chat2Gen.createMetaRequestTrusted({
+            conversationIDKeys: [conversationIDKey],
+            reason: 'ensureChannelMeta',
+          })
+        )
+      }
+      dispatch(TeamsGen.createGetMembers({teamID}))
+      dispatch(UsersGen.createGetBlockState({usernames: participants}))
+    }
+  }, [
+    selectedTab,
+    dispatch,
+    conversationIDKey,
+    prevSelectedTab,
+    meta.conversationIDKey,
+    participants,
+    teamID,
+  ])
+  React.useEffect(() => {
+    if (selectedTab !== prevSelectedTab && selectedTab === 'bots') {
+      // Load any bots that aren't in the featured bots map already
+      bots
+        .filter(botUsername => !featuredBotsMap.has(botUsername))
+        .map(botUsername => dispatch(BotsGen.createSearchFeaturedBots({query: botUsername})))
+    }
+  }, [selectedTab, dispatch, conversationIDKey, prevSelectedTab, bots, featuredBotsMap])
 }
 
-const Channel = (props: Props & TabProps) => {
-  const {teamID, conversationIDKey, selectedTab, setSelectedTab} = props
-  const renderItem = ({item}: {item: Row}) => {
-    switch (item.type) {
-      case 'tabs':
-        return (
-          <ChannelTabs
-            teamID={teamID}
+// keep track during session
+const lastSelectedTabs: {[T: string]: TabKey} = {}
+const defaultTab: TabKey = 'members'
+
+// TODO: consider using this for the teams/team/tabs state code
+const useTabsState = (
+  conversationIDKey: ChatTypes.ConversationIDKey,
+  providedTab?: TabKey
+): [TabKey, (t: TabKey) => void] => {
+  const defaultSelectedTab = lastSelectedTabs[conversationIDKey] ?? providedTab ?? defaultTab
+  const [selectedTab, _setSelectedTab] = React.useState<TabKey>(defaultSelectedTab)
+  const setSelectedTab = React.useCallback(
+    t => {
+      lastSelectedTabs[conversationIDKey] = t
+      _setSelectedTab(t)
+    },
+    [conversationIDKey, _setSelectedTab]
+  )
+
+  const prevConvID = Container.usePrevious(conversationIDKey)
+
+  React.useEffect(() => {
+    if (conversationIDKey !== prevConvID) {
+      setSelectedTab(defaultSelectedTab)
+    }
+  }, [conversationIDKey, prevConvID, setSelectedTab, defaultSelectedTab])
+  return [selectedTab, setSelectedTab]
+}
+
+const emptyMap = new Map()
```

Comment:
Why is this declared out here? If we use it twice, the two uses would reference the same map.

Reply:
Exactly. This way, when `Container.useSelector(state => state.teams.teamIDToMembers.get(teamID) ?? emptyMap)` returns it twice, they aren't mistaken for different things, triggering a pointless rerender.

Reply:
In that case, please name it so it's only ever used in this particular spot.